multicore_runtime/middleware.py

```diff
@@ -0,0 +1,193 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""WSGI middleware to wrap the dispatcher, and supporting functions."""
+
+import httplib
+import logging
+import os
+
+from werkzeug.wrappers import Request
+from werkzeug.wrappers import Response
+
+
+# A dict of reserved env keys; the value is used as the default if not
+# otherwise set.
+RESERVED_ENV_KEYS = {
+    'AUTH_DOMAIN': 'gmail.com',  # Default auth domain must be set.
+    'DATACENTER': '',
+    'DEFAULT_VERSION_HOSTNAME': '',
+    'HTTPS': '',
+    'REMOTE_ADDR': '',
+    'REQUEST_ID_HASH': '',
+    'REQUEST_LOG_ID': '',
+    'USER_EMAIL': '',
+    'USER_ID': '',
+    'USER_IS_ADMIN': '0',  # Default admin flag to explicit '0'.
+    'USER_NICKNAME': '',
+    'USER_ORGANIZATION': '',
+    }
+
+
+def reset_environment_middleware(app, frozen_environment, frozen_user_env,
+                                 frozen_env_config_env):
+  """Replace the contents of os.environ with a frozen env + request data.
+
+  This requires a single-threaded webserver, or for os.environ to be patched to
+  be thread-local.
+
+  Args:
+    app: The WSGI app to wrap.
+    frozen_environment: An iterable of (key, value) tuples that can be used to
+      populate os.environ with the initial state of the environment.
+      `tuple(os.environ.iteritems())` produces appropriate output.
+    frozen_user_env: An iterable of (key, value) tuples that can be used to
+      populate os.environ with environment variables specified in app.yaml.
+    frozen_env_config_env: An iterable of (key, value) tuples that can be
+      used to populate os.environ with configuration-dependent env variables.
+
+  Returns:
+    The wrapped app, also a WSGI app.
+  """
+
+  def reset_environment_wrapper(wsgi_env, start_response):
```

Comment:
Why not use `@Request.application` (or `@responder`) here too?

I think you should try to be as uniform as possible in your use of it. If it can be used everywhere, why not to?